File: go/src/383.赎金信.go
/*
 * @lc app=leetcode.cn id=383 lang=golang
 *
 * [383] 赎金信
 */

// @lc code=start
func canConstruct(ransomNote string, magazine string) bool {
	letters := make([]int, 26)
	for i := 0; i < len(magazine); i++ {
		letters[magazine[i] - 'a']++
	}

	for i := 0; i < len(ransomNote); i++ {
		if letters[ransomNote[i] - 'a'] <= 0 {
			return false
		}
		letters[ransomNote[i] - 'a']--
	}

	return true
}
// @lc code=end

File: rust/src/383.赎金信.rs
/*
 * @lc app=leetcode.cn id=383 lang=rust
 *
 * [383] 赎金信
 */

// @lc code=start
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut counter = vec![0; 26];
        magazine.as_bytes().iter().for_each(|&c| {
            counter[(c - 'a' as u8) as usize] += 1;
        });

        let bytes = ransom_note.as_bytes();
        for i in 0..bytes.len() {
            let index = (bytes[i] - 'a' as u8) as usize;
            counter[index] -= 1;
            if counter[index] < 0 {
                return false;
            }
        }
        true
    }
}
// @lc code=end
